File: src/ht.c
#include "ht.h"

static inline void set_tree_address(uint32_t adrs[ADRS_LEN], idx_tree_t idx) {
  idx_tree_t tmp;
  tmp.valhb = htobe32(idx.valhb);
  tmp.vallb = htobe64(idx.vallb);

  memcpy(&adrs[ADRS_TREE_ADDRESS_IDX], &tmp, sizeof(tmp));
}

/*
 * Generate a hypertree signature.
 */
int ht_sign(uint8_t *out, const uint8_t msg[ENN], const uint8_t sk_seed[ENN],
            const uint8_t pk_seed[ENN], idx_tree_t idx_tree,
            uint64_t idx_leaf) {
  uint32_t adrs[ADRS_LEN] = {0};
  set_tree_address(adrs, idx_tree);

  uint8_t *csig = out;
  xmss_sign(csig, msg, sk_seed, idx_leaf, pk_seed, adrs);

  uint8_t root[ENN];
  xmss_pk_from_sig(root, idx_leaf, csig, msg, pk_seed, adrs);

  csig += (WOTSP_LEN + HP) * ENN;
  size_t j;
  for (j = 1; j < D; j++) {
    idx_leaf = idx_tree_mod_2hp(idx_tree);
    idx_tree_shift(&idx_tree, HP);

    adrs[ADRS_LAYER_ADDRESS_IDX] = htobe32(j);
    set_tree_address(adrs, idx_tree);

    xmss_sign(csig, root, sk_seed, idx_leaf, pk_seed, adrs);

    if (j < D - 1) {
      xmss_pk_from_sig(root, idx_leaf, csig, root, pk_seed, adrs);
    }

    csig += (WOTSP_LEN + HP) * ENN;
  }

  return 0;
}

int ht_verify(const uint8_t msg[ENN], const uint8_t *sig_ht,
              const uint8_t pk_seed[ENN], idx_tree_t idx_tree,
              uint64_t idx_leaf, const uint8_t pk_root[ENN]) {
  uint32_t adrs[ADRS_LEN] = {0};
  set_tree_address(adrs, idx_tree);

  const uint8_t *csig = sig_ht;
  uint8_t node[ENN];
  xmss_pk_from_sig(node, idx_leaf, csig, msg, pk_seed, adrs);
  csig += (WOTSP_LEN + HP) * ENN;

  size_t j;
  for (j = 1; j < D; j++) {
    idx_leaf = idx_tree_mod_2hp(idx_tree);
    idx_tree_shift(&idx_tree, HP);

    adrs[ADRS_LAYER_ADDRESS_IDX] = htobe32(j);
    set_tree_address(adrs, idx_tree);

    xmss_pk_from_sig(node, idx_leaf, csig, node, pk_seed, adrs);
    csig += (WOTSP_LEN + HP) * ENN;
  }

  if (memcmp(node, pk_root, ENN) == 0)
    return 1;
  else
    return 0;
}

File: python/ht.py
"""
From Section 7.
Hypertree section.
"""

from address import Address
from slh_dsa import SLHDSA
from xmss import xmss_sign, xmss_pk_from_sig
from utils import toByte

def ht_sign(M: bytes, sk_seed: bytes, pk_seed: bytes, idx_tree: int, idx_leaf: int, ctx: SLHDSA) -> bytes:
    """Generate a hypertree signature"""
    assert idx_leaf >= 0
    assert idx_tree >= 0

    adrs = Address(toByte(0,32)) # adrs <- toByte(0, 32)

    adrs.set_tree_address(idx_tree)
    sig_tmp = xmss_sign(M, sk_seed, idx_leaf, pk_seed, adrs, ctx)
    sig_ht = bytes(sig_tmp)
    root = xmss_pk_from_sig(idx_leaf, sig_tmp, M, pk_seed, adrs, ctx)
    for j in range(1, ctx.d):
        idx_leaf = idx_tree % (2**ctx.hp)
        idx_tree = idx_tree >> ctx.hp
        adrs.set_layer_address(j)
        adrs.set_tree_address(idx_tree)
        sig_tmp = xmss_sign(root, sk_seed, idx_leaf, pk_seed, adrs, ctx)
        sig_ht += sig_tmp
        if j < ctx.d - 1:
            root = xmss_pk_from_sig(idx_leaf, sig_tmp, root, pk_seed, adrs, ctx)
    return sig_ht

def ht_verify(M: bytes, sig_ht: bytes, pk_seed: bytes, idx_tree: int, idx_leaf: int, pk_root: bytes, ctx: SLHDSA) -> bool:
    """Verify a hypertree signature"""
    assert idx_leaf >= 0
    assert idx_tree >= 0

    adrs = Address(toByte(0,32)) # adrs <- toByte(0, 32)

    adrs.set_tree_address(idx_tree)
    sig_tmp = sig_ht[:(ctx.hp+ctx.wotsp_len)*ctx.n]  # sig_tmp <- sig_ht.getXMSSSignature(0)
    node = xmss_pk_from_sig(idx_leaf, sig_tmp, M, pk_seed, adrs, ctx)
    for j in range(1, ctx.d):
        idx_leaf = idx_tree % (2**ctx.hp)
        idx_tree = idx_tree >> ctx.hp
        adrs.set_layer_address(j)
        adrs.set_tree_address(idx_tree)
        # sig_tmp <- sig_ht.getXMSSSignature(j)
        sig_tmp = sig_ht[j*(ctx.hp+ctx.wotsp_len)*ctx.n : (j+1)*(ctx.hp+ctx.wotsp_len)*ctx.n]
        node = xmss_pk_from_sig(idx_leaf, sig_tmp, node, pk_seed, adrs, ctx)
    if node == pk_root:
        return True
    else:
        return False